### ISSUE TITLE
malloc: Incorrect checksum for freed object

### ISSUE THREAD

**Issue by Programmierus** [alex-petrenko/faster-fifo#31]
I am trying to replace `multiprocessing.Queue` in three different places of my code and for some reason one of them gives me following errors and doesn't work:

```
python(27304,0x70001d443000) malloc: Incorrect checksum for freed object 0x7fa22107ea00: probably modified after being freed.
Corrupt value: 0x2c4
python(27304,0x70001d443000) malloc: *** set a breakpoint in malloc_error_break to debug
```

I am trying to research further, but so far could not think about reasonable direction :) It works with standard queue without issues (all I am trying to achieve is a speed gain)

**UPDATE** It seems to have something to do with a queue being empty. If I add a check before `get()`:

```
if self.data_queue.empty():
    continue
```

It works. I am also getting `AssertionError` sometimes on `get()` which I ignore and continue at the moment...

**Comment by alex-petrenko** [comment]
Thank you for reporting this! Looks like this is the first time someone encountered this problem.
Any chance you can isolate the issue and provide some code that reproduces it? Also, what platform are you on?

What kinds of objects are you putting into the queue? If they're using custom serialization of any kind, this can potentially be a place to look at.

Additionally, one thing I'd try is running the app with Valgrind memcheck to see if there are any warnings.

If the `empty()` check helps for now this indeed indicates the problem might happen when you're trying to get() from an empty queue. This check is not bulletproof BTW, in a multi-process app the queue might become empty immediately after you checked. Although this check probably massively reduces the probability of this issue reproducing. If you reduce the size of the circular buffer when the queue is constructed, does this trigger the error faster? (in case it is something related to circular buffer wrapping around)

More info on that AssertionError would be helpful too, i.e. where it happens and if there's any message attached to it.

**Comment by alex-petrenko** [comment]
The reason I'm inquiring about serializers/deserializers is because the queue code itself does not allocate or deallocate any memory: https://github.com/alex-petrenko/faster-fifo/blob/master/cpp_faster_fifo/cpp_lib/faster_fifo.cpp
It even initializes the Queue object itself in memory provided by Python using placement new.

But the problem can still be caused by the C++ code writing somewhere past the buffer boundaries, triggering some kind of UB.

**Comment by Programmierus** [comment]
Hello, firstly thanks for your detailed reaction on my issue. Thumbs up! 

Currently I have stepped of a bit from faster_fifo due to this problem and working on my project without it.

I have moved the code producing mentioned errors to a branch and will get back to it with 3.3.1 in debug mode to collect more information and will get back here.

**Comment by alex-petrenko** [comment]
If you can, try latest version 1.4.1
We fixed some issues related to multithreading, so if you ever used faster-fifo with threads and not exclusively processes, this might fix your issue.

**Comment by Programmierus** [comment]
I am sorry I left this issue behind :( I completely reworked the code base of the mentioned project and the problem was gone. The branch that had this problem is not restorable :( Closing for now.

**Comment by alex-petrenko** [comment]
Thank you for updating!
I'm glad it worked out!

There's a good chance that recent fixes in 1.4.1 and 1.4.2 helped. I am using this in multiple projects and things have been very stable so far.
